internal/restic/map.go

```diff
@@ -0,0 +1,186 @@
+package restic
+
+import "hash/maphash"
+
+// A Map is a mapping of BlobHandle to V.
+// There may be multiple entries for a given handle.
+type Map[V any] struct {
+	byType [NumBlobTypes]idmap[V]
+
+	mh maphash.Hash
+
+	free *MapEntry[V] // Free list.
+}
+
+// An idmap is a chaining hash table that maps blob IDs to values of type V.
+// It allows storing multiple entries with the same key.
+//
+// Map uses some optimizations that are not compatible with supporting
+// deletions.
+//
+// The buckets in this hash table contain only pointers, rather than inlined
+// key-value pairs like the standard Go map. This way, only a pointer array
+// needs to be resized when the table grows, preventing memory usage spikes.
+type idmap[V any] struct {
+	// The number of buckets is always a power of two and never zero.
+	buckets    []*MapEntry[V]
+	numentries uint
+}
+
+// A MapEntry is an entry in a Map with an ID and a value V.
+// There may be multiple entries for a given ID.
+type MapEntry[V any] struct {
+	ID   ID // Key.
+	Data V  // Satellite data. Not touched by Map.
+	next *MapEntry[V]
+}
+
+const (
+	growthFactor = 2 // Must be a power of 2.
+	maxLoad      = 4 // Max. number of entries per bucket.
+)
+
+// Add creates a new MapEntry for h in the map and returns it.
+func (m *Map[V]) Add(h BlobHandle) *MapEntry[V] {
+	tm := &m.byType[h.Type]
+
+	switch {
+	case tm.numentries == 0: // Lazy initialization.
+		tm.init()
+	case tm.numentries >= maxLoad*uint(len(tm.buckets)):
+		m.grow(tm)
+	}
+
+	hash := tm.hindex(m.hash(h.ID))
+	e := m.newEntry()
+	e.ID = h.ID
+	e.next = tm.buckets[hash] // Prepend to existing chain.
+
+	tm.buckets[hash] = e
+	tm.numentries++
+
+	return e
+}
+
+// Foreach calls fn for all entries in the map, until fn returns false.
+// Callers may modify a MapEntry's V, but not its ID.
+func (m *Map[V]) Foreach(fn func(*MapEntry[V], BlobType) bool) {
+	for typ := range m.byType {
+		for _, e := range m.byType[typ].buckets {
+			for e != nil {
+				if !fn(e, BlobType(typ)) {
+					return
+				}
+				e = e.next
+			}
+		}
+	}
+}
+
+// ForeachWithID calls fn for all entries with the given h.
+// Callers may modify an entry's V, but not its ID.
+func (m *Map[V]) ForeachWithID(h BlobHandle, fn func(*MapEntry[V])) {
+	tm := &m.byType[h.Type]
+	if len(tm.buckets) == 0 {
+		return
+	}
+
+	hash := tm.hindex(m.hash(h.ID))
+	for e := tm.buckets[hash]; e != nil; e = e.next {
+		if e.ID != h.ID {
+			continue
+		}
+		fn(e)
+	}
+}
+
+// Get returns the first entry for the given id.
+func (m *Map[V]) Get(h BlobHandle) *MapEntry[V] {
+	tm := m.byType[h.Type]
```

Comment:
That creates a copy of the map on the stack. Not sure though whether that is the cause of the performance problems.

Reply:
Fixed. This is a part of it, but it's not the whole story.

```
name                                             old time/op  new time/op  delta
IndexHasUnknown-8                                47.4ns ±16%  44.4ns ±11%     ~     (p=0.371 n=10+5)
MasterIndexLookupSingleIndexUnknown-8            70.1ns ± 9%  76.8ns ±11%     ~     (p=0.165 n=10+5)
MasterIndexLookupMultipleIndexUnknown-8          70.8ns ± 8%  70.4ns ± 7%     ~     (p=0.768 n=10+5)
MasterIndexLookupParallel/unknown,indices=25-8    382ns ± 2%   321ns ± 2%  -15.99%  (p=0.001 n=10+5)
MasterIndexLookupParallel/unknown,indices=50-8    369ns ± 8%   359ns ± 3%     ~     (p=0.165 n=10+5)
MasterIndexLookupParallel/unknown,indices=100-8   392ns ± 3%   364ns ± 0%   -7.08%  (p=0.002 n=10+4)
MasterIndexLookupBlobSize-8                      63.0ns ± 9%  64.3ns ± 9%     ~     (p=0.594 n=10+5)
```